tests/scenario/test_ingress_per_app.py

```diff
@@ -0,0 +1,162 @@
+def test_ingress_per_app_created(
+    traefik_ctx, port, host, model, traefik_container, event_name, tmp_path
+):
+    """Check the config when a new ingress per leader is created or changes (single remote unit)."""
+    ipa = Relation(
+        "ingress",
+        remote_app_data={
+            "model": "test-model",
+            "name": "remote/0",
+            "mode": "http",
+        },
+        remote_units_data={0: {"port": str(port), "host": host}},
+        relation_id=1,
+    )
+    state = State(
+        model=model,
+        config={"routing_mode": "path", "external_hostname": "foo.com"},
+        containers=[traefik_container],
+        relations=[ipa],
+    )
+
+    # WHEN any relevant event fires
+    event = getattr(ipa, f"{event_name}_event")
+    traefik_ctx.run(event, state)
+
+    generated_config = yaml.safe_load(
+        traefik_container.filesystem.open(
+            f"/opt/traefik/juju/juju_ingress_ingress_{ipa.relation_id}_remote.yaml"
+        ).read()
+    )
+
+    assert generated_config["http"]["services"]["juju-test-model-remote-0-service"] == {
+        "loadBalancer": {"servers": [{"url": f"http://{host}:{port}"}]}
+    }
+
+
+@pytest.mark.parametrize("port, host", ((80, "1.1.1.{}"), (81, "10.1.10.{}")))
+@pytest.mark.parametrize("n_units", (2, 3, 10))
+def test_ingress_per_app_scale(
+    traefik_ctx, host, port, model, traefik_container, tmp_path, n_units
+):
+    """Check the config when a new ingress per leader unit joins."""
+    cfg_file = tmp_path.joinpath("traefik", "juju", "juju_ingress_ingress_1_remote.yaml")
+    cfg_file.parent.mkdir(parents=True)
+
+    # config that would have been generated from mock_data_0
+    # same as config output of the previous test
+    initial_cfg = {
+        "http": {
+            "routers": {
+                "juju-test-model-remote-0-router": {
+                    "entryPoints": ["web"],
+                    "rule": "PathPrefix(`/test-model-remote-0`)",
+                    "service": "juju-test-model-remote-0-service",
+                },
+                "juju-test-model-remote-0-router-tls": {
+                    "entryPoints": ["websecure"],
+                    "rule": "PathPrefix(`/test-model-remote-0`)",
+                    "service": "juju-test-model-remote-0-service",
+                    "tls": {"domains": [{"main": "foo.com", "sans": ["*.foo.com"]}]},
+                },
+            },
+            "services": {
+                "juju-test-model-remote-0-service": {
+                    "loadBalancer": {"servers": [{"url": f"http://{host}:{port}"}]}
+                }
+            },
+        }
+    }
+    cfg_file.write_text(yaml.safe_dump(initial_cfg))
+
+    def _get_mock_data(n: int):
+        return {
+            "port": str(port + n),
+            "host": host.format(n),
+        }
+
+    ipa = Relation(
+        "ingress",
+        remote_app_data={
+            "model": "test-model",
+            "name": "remote/0",
```

Comment:
I'm a bit confused:
1. Who is being scaled here - traefik or the app asking for an ingress url?
2. Why app name has a unit index? And why was it converted to dash by IPA (`remote-0`)

Reply:
1. the app asking for an ingress url is being scaled: you can see that from remote_units_data being populated by `n` databags
2. very sharply spotted. that's a leftover from a IPU test probably. It was converted to a dash because IPA and IPU prefixes are handled the same, and charm.TraefikIngressCharm._get_prefix converts slash to dash to normalize unit names.